### PR TITLE
Test

### PR DESCRIPTION
**Pull request by irisliu77**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +270,104 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void optionalAttendeeWithWholedayEvent() {
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:275]: For your tests, it'd be nice to add comments similar to the other tests to describe what you're testing for.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -111,7 +111,7 @@ public void everyAttendeeIsConsidered() {
 
     MeetingRequest request =
         new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
-
+
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:114]: Nit: remove whitespace

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,99 @@
...
+      }
+    }
+    return res;
+  }
+
+  private List<TimeRange> getAvailableTime(List<TimeRange> timeRanges, long duration) {
+    List<TimeRange> res = new ArrayList<>();
+    int prevEnd = TimeRange.START_OF_DAY;
+    for (int i = 0; i < timeRanges.size(); i++) {
+      TimeRange avaialbeTimeRange = TimeRange.fromStartEnd(prevEnd, timeRanges.get(i).start(), false);
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:99]: Nit: spelling "available" instead of "avaialbe"

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,99 @@
...
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // if the request meeting is longer than the duration of one day
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: Nit: for comments, can you capitalize the first word and add a period at the end? aka make them complete sentences.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,99 @@
...
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    // if no mandatory attendees or optional attendees, return whole day
+    if (mandatoryAttendees.size() == 0 && optionalAttendees.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // get unavailable time of mandatory attendees
+    List<TimeRange> unavailableTime = getUnavailableTime(events, mandatoryAttendees);
+    // merge available intervals with overlaps
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:44]: Nit: I think this comment should be merge **unavailable** intervals

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,99 @@
...
+    List<TimeRange> res = getAvailableTime(mergedUnavailableTime, request.getDuration());
+    return res;
+  }
+
+  private List<TimeRange> getUnavailableTime(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> res = new ArrayList<>();
+    for (Event e : events) {
+      Set<String> attendees = e.getAttendees();
+      for (String attendee : attendees) {
+        // if there is a least on mandatory attendee for the requested event need to attend this event
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:56]: Nit: "If there is at least one mandatory attendee for the requested event, mark the duration of this event as unavailable."

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,99 @@
...
+    List<TimeRange> unavailableTime = getUnavailableTime(events, mandatoryAttendees);
+    // merge available intervals with overlaps
+    List<TimeRange> mergedUnavailableTime = mergeUnavailableTime(unavailableTime);
+    // get available time slot base on merged unavailable time
+    List<TimeRange> res = getAvailableTime(mergedUnavailableTime, request.getDuration());
+    return res;
+  }
+
+  private List<TimeRange> getUnavailableTime(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> res = new ArrayList<>();
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:52]: Nit: change `res` to be something like `unavailableTimes` so it's more descriptive.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,99 @@
...
+          break;
+        }
+      }
+    }
+    return res;
+  }
+
+  private List<TimeRange> mergeUnavailableTime(List<TimeRange> timeRanges) {
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> res = new ArrayList<>();
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:69]: Nit: change to be `mergedUnavailableTimes` or something along those lines.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,99 @@
...
+
+    for (TimeRange t : timeRanges) {
+      // if the result list is empty, we can add it to the result list directly
+      if (res.isEmpty()) {
+        res.add(t);
+      } else {
+        TimeRange latestTimeRange = res.get(res.size() - 1);
+
+        // if the latest time slot that has been processed doesn't overlap with the time slot currently being processed
+        if (!t.overlaps(latestTimeRange)) {
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:79]: I would move this comment after the if check so it's clear that it's referring to the case where there is no overlap.

Maybe reword to something like: `If the previous time slot doesn't overlap with the current time slot, we can add it to the result list directly`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,99 @@
...
+          res.remove(res.size() - 1);
+          res.add(TimeRange.fromStartEnd(latestTimeRange.start(), end, false));
+        }
+      }
+    }
+    return res;
+  }
+
+  private List<TimeRange> getAvailableTime(List<TimeRange> timeRanges, long duration) {
+    List<TimeRange> res = new ArrayList<>();
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:96]: Rename to availableTimes or something like that

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,99 @@
...
+          // remove the previous time range with overlap
+          res.remove(res.size() - 1);
+          res.add(TimeRange.fromStartEnd(latestTimeRange.start(), end, false));
+        }
+      }
+    }
+    return res;
+  }
+
+  private List<TimeRange> getAvailableTime(List<TimeRange> timeRanges, long duration) {
```
Comment (joanchenx) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:95]: Nit: `getAvailableTimes` instead (so plural since you're returning a list)

**Comment by irisliu77** [comment]
Ready for review!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,97 @@
...
+      }
+    }
+    return unavailableTimes;
+  }
+
+  private List<TimeRange> mergeUnavailableTimes(List<TimeRange> timeRanges) {
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    List<TimeRange> mergedUnavailableTimes = new ArrayList<>();
+
+    for (TimeRange t : timeRanges) {
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:69]: To understand this algorithm, I think the important missing invariant (necessary to understand why the code is the way it is) is that `mergedUnavailableTimes` is a list of sorted, non-overlapping `TimeRange`s. Should that be documented?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -14,10 +14,97 @@
...
+          // we can add it to the merged list directly.
+          mergedUnavailableTimes.add(t);
+        } else {
+          // Since we sort the time range by start time,
+          // the start time of previous events should be earlier or equal to current event,
+          // and we keep the later end time.
+          int end = Math.max(t.end(), latestTimeRange.end());
+          // Remove the previous time range with overlap.
+          mergedUnavailableTimes.remove(mergedUnavailableTimes.size() - 1);
+          mergedUnavailableTimes.add(TimeRange.fromStartEnd(latestTimeRange.start(), end, false));
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:86]: I think you just want to use `.set` here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,97 @@
...
+    // Get available time slot base on merged unavailable time.
+    List<TimeRange> availableTimes = getAvailableTimes(mergedUnavailableTimes, request.getDuration());
+    return availableTimes;
+  }
+
+  private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    for (Event e : events) {
+      Set<String> attendees = e.getAttendees();
+      for (String attendee : attendees) {
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: Sounds like you're looking for a set intersection. Maybe that would be a less verbose way of doing this?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -14,10 +14,97 @@
...
+        }
+      }
+    }
+    return mergedUnavailableTimes;
+  }
+
+  private List<TimeRange> getAvailableTimes(List<TimeRange> timeRanges, long duration) {
+    List<TimeRange> availableTimes = new ArrayList<>();
+    int prevEnd = TimeRange.START_OF_DAY;
+    for (int i = 0; i < timeRanges.size(); i++) {
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:96]: Given that you only ever access the ith element, you can use a for-each loop instead to avoid having an extra variable `i`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,97 @@
...
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // If there is no event on the calendar currently, return whole day as possible time slot.
+    if (events.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: Unused variable.

Comment (anbenson) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: Well, you essentially don't deal with optional attendees.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,97 @@
...
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // If the request meeting is longer than the duration of one day, return an empty list.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // If there is no event on the calendar currently, return whole day as possible time slot.
+    if (events.size() == 0) {
```
[the remark this answers is not in the document]
Comment (anbenson) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:30]: (and later on)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 18]
```diff
@@ -270,5 +269,120 @@ public void notEnoughRoom() {
...
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_0800AM, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_B)));
+    
+    MeetingRequest request = new MeetingRequest(Arrays.asList(), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:384]: Collections.EMPTY_LIST

**Comment by irisliu77** [comment]
I have implemented algorithm for optional attendees and all the tests work now

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -14,10 +14,119 @@
...
+    // Merge unavailable intervals with overlaps.
+    List<TimeRange> mergedUnavailableTimes = mergeUnavailableTimes(unavailableTimes);
+    // Get available time slot base on merged unavailable time.
+    List<TimeRange> availableTimes = getAvailableTimes(mergedUnavailableTimes, request.getDuration());
+    // Deal with optional attendees.
+    if (!optionalAttendees.isEmpty()) {
+      List<TimeRange> optionalUnavailableTimes = getUnavailableTimes(events, optionalAttendees);
+      optionalUnavailableTimes.addAll(mergedUnavailableTimes);
+      List<TimeRange> optionalMergedUnavailableTimes = mergeUnavailableTimes(optionalUnavailableTimes);
+      List<TimeRange> combineAvailableTimes = getAvailableTimes(optionalMergedUnavailableTimes, request.getDuration());
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:51]: In these ~10 lines, there are 9 symbols containing the substring "available" used a combined total of 18 times.
- unavailableTimes
- getUnavailableTimes
- mergedUnavailableTimes
- mergeUnavailableTimes
- availableTimes
- optionalUnavailableTimes
- optionalMergedUnavailableTimes
- combineAvailableTimes

I think the naming here requires readers to spend a lot of energy focusing and reading multiple times to understand what's happening here.

I mainly want to point it out as food for thought and am not recommending a particular solution. But for example, I think this reads a little easier and is a bit more accurate as to what the methods do:
```java
List<TimeRange> unavailableTimes = findBusyTimes(events, mandatoryAttendees);
unavailableTimes = mergeOverlappingTimes(unavailableTimes);
List<TimeRange> availableTimes = findAvailableTimes(unavailableTimes, request.getDuration());
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -14,10 +14,119 @@
...
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Get unavailable time of mandatory attendees.
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, mandatoryAttendees);
+    // Merge unavailable intervals with overlaps.
+    List<TimeRange> mergedUnavailableTimes = mergeUnavailableTimes(unavailableTimes);
+    // Get available time slot base on merged unavailable time.
+    List<TimeRange> availableTimes = getAvailableTimes(mergedUnavailableTimes, request.getDuration());
+    // Deal with optional attendees.
+    if (!optionalAttendees.isEmpty()) {
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: I know Joan asked you to write comments, but these comments do not add value. They mostly merely say the same thing as the following line. `// Get unavailable time of mandatory attendees` and `getUnavailableTimes(events, mandatoryAttendees);` are basically the same thing.

I like it when comments explain the higher-level picture of what a block of code is trying to accomplish - for example, the overall algorithm you're implementing in this method could be explained.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -14,10 +14,119 @@
...
+        return combineAvailableTimes;
+      }
+    }
+    return availableTimes;
+  }
+
+  /**
+  * @param  events  a list of existing events.
+  * @param  mandatoryAttendees a list of mandatory attendees for the requested event.
+  * @return a list of maybe overlaping timeRange that at least one mandatory attendee is unavailable
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:64]: overlapping

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,119 @@
...
+      if (mandatoryAttendees.isEmpty()) {
+        return combineAvailableTimes;
+      } else if (!combineAvailableTimes.isEmpty()) {
+        return combineAvailableTimes;
+      }
+    }
+    return availableTimes;
+  }
+
+  /**
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:61]: If I were really nitpicky about Javadoc...the spaces aren't aligned on each line e.g. exactly one space between @ param and the next word. Punctuation (a period) does not follow the sentence fragment unless there is a following sentence.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -14,10 +14,119 @@
...
+  * @param  events  a list of existing events.
+  * @param  mandatoryAttendees a list of mandatory attendees for the requested event.
+  * @return a list of maybe overlaping timeRange that at least one mandatory attendee is unavailable
+  */
+  private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> mandatoryAttendees) {
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    for (Event e : events) {
+      Set<String> attendees = e.getAttendees();
+      Set<String> intersection = new HashSet<String>(attendees);
+      intersection.retainAll(mandatoryAttendees);
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: There's no need to turn these into sets. Collections.disjoint can work on Collections instead of requiring Sets.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -14,10 +14,119 @@
...
+          // If the previous time slot doesn't overlap with the current time slot, 
+          // we can add it to the merged list directly.
+          mergedUnavailableTimes.add(t);
+        } else {
+          // Since we sort the time range by start time,
+          // the start time of previous events should be earlier or equal to current event,
+          // and we keep the later end time.
+          int end = Math.max(t.end(), latestTimeRange.end());
+          // Set the previous time range to be the merged new time range.
+          mergedUnavailableTimes.set(mergedUnavailableTimes.size() - 1, TimeRange.fromStartEnd(latestTimeRange.start(), end, false));
```
Comment (anbenson) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:103]: Split up super long lines into different lines.
